spring-cloud-gcp-logging/src/main/java/org/springframework/cloud/gcp/logging/StackdriverJsonLayout.java

```diff
@@ -0,0 +1,229 @@
+	 */
+	public void setProjectId(String projectId) {
+		this.projectId = projectId;
+	}
+
+	/**
+	 *
+	 * @return true if the traceId should be included into the json
+	 */
+	public boolean isIncludeTraceId() {
+		return this.includeTraceId;
+	}
+
+	/**
+	 * @param includeTraceId true if the traceId should be included into the json
+	 */
+	public void setIncludeTraceId(boolean includeTraceId) {
+		this.includeTraceId = includeTraceId;
+	}
+
+	/**
+	 * @return true if the spanId should be included into the json
+	 */
+	public boolean isIncludeSpanId() {
+		return this.includeSpanId;
+	}
+
+	/**
+	 * @param includeSpanId true if the spanId should be included into the json
+	 */
+	public void setIncludeSpanId(boolean includeSpanId) {
+		this.includeSpanId = includeSpanId;
+	}
+
+	/**
+	 *
+	 * @return true if the exception should be added to the message
+	 */
+	public boolean isIncludeExceptionInMessage() {
+		return this.includeExceptionInMessage;
+	}
+
+	/**
+	 * @param includeExceptionInMessage true if the exception should be added to the message
+	 */
+	public void setIncludeExceptionInMessage(boolean includeExceptionInMessage) {
+		this.includeExceptionInMessage = includeExceptionInMessage;
+	}
+
+	/**
+	 * @param event the logging event
+	 * @return the map which should get rendered as json
+	 */
+	@Override
+	protected Map<String, Object> toJsonMap(ILoggingEvent event) {
+
+		Map<String, Object> map = new LinkedHashMap<>();
+
+		if (this.includeMDC) {
+			event.getMDCPropertyMap().forEach((key, value) -> {
+				if (FILTERED_MDC_FIELDS.contains(key)) {
+					return;
+				}
+				map.put(key, value);
+			});
+		}
+		if (this.includeTimestamp) {
+			map.put(TIMESTAMP_SECONDS_ATTRIBUTE, TimeUnit.MILLISECONDS.toSeconds(event.getTimeStamp()));
+			map.put(TIMESTAMP_NANOS_ATTRIBUTE, TimeUnit.MILLISECONDS.toNanos(event.getTimeStamp() % 1_000));
+		}
+
+		add(SEVERITY_ATTRIBUTE, this.includeLevel, String.valueOf(event.getLevel()), map);
+		add(THREAD_ATTR_NAME, this.includeThreadName, event.getThreadName(), map);
+		add(LOGGER_ATTR_NAME, this.includeLoggerName, event.getLoggerName(), map);
+
+		if (this.includeFormattedMessage) {
+			String message = event.getFormattedMessage();
+			if (this.includeExceptionInMessage) {
+				IThrowableProxy throwableProxy = event.getThrowableProxy();
+				if (throwableProxy != null) {
+					String stackTrace = getThrowableProxyConverter().convert(event);
+					if (stackTrace != null && !stackTrace.equals("")) {
+						message += "\n" + stackTrace;
+					}
+				}
+			}
+			map.put(FORMATTED_MESSAGE_ATTR_NAME, message);
+		}
+		add(MESSAGE_ATTR_NAME, this.includeMessage, event.getMessage(), map);
+		add(CONTEXT_ATTR_NAME, this.includeContextName, event.getLoggerContextVO().getName(), map);
+		addThrowableInfo(EXCEPTION_ATTR_NAME, this.includeException, event, map);
+		addTraceId(event, map);
+		add(SPAN_ID_ATTRIBUTE, this.includeSpanId, event.getMDCPropertyMap().get(MDC_FIELD_SPAN_ID), map);
```

Comment:
How is the MDC span ID field set?

For Stackdriver Trace, the LogEntry `span_id` field should be hexadecimal encoded for the Stackdriver Trace viewer to know where to display the log entries as you look at a trace, per the [Trace Viewer > Log Entries doc](https://cloud.google.com/trace/docs/viewing-details#log_entries). This is because span IDs are specified as hex-encoded strings in the Stackdriver Trace API v2, see the [Span doc](https://cloud.google.com/trace/docs/reference/v2/rpc/google.devtools.cloudtrace.v2#span).

My thinking is that this should affect how the field is added to the MDC so this could would be fine as-is.

Reply:
@draffensperger
Only the `traceId` is currently being processed. The connection of GCP-Traces and GCP-Logs works perfectly here. I only added the `spanid` because it's also [parsed by the fluentd sidecar](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/blob/master/lib/fluent/plugin/out_google_cloud.rb#L123). Here I could not find any usage in the stackdriver at all.

Reply:
OK, makes sense. Yes, the support for `span_id` in the Stackdriver Trace viewer is quite new. I think it makes sense to keep it here as yes, it's being parsed by the fluentd sidecar, but we can get it into the MDC late on.

Reply:
Just for clarification: the `spanId` is already set by spring sleuth in the MDC and evaluated correctly by this `StackdriverJsonLayout`.

Reply:
The logging module may be used without Sleuth, and the `traceId` should be extracted from `TraceLoggingEnhancer.getCurrentTraceId()`. We may be able to put it into the MDC, but currently it's not happening.
See `TraceIdLoggingWebMvcInterceptor`.